src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -5339,23 +5445,23 @@ static synchronized long getColumnEncryptionKeyCacheTtl() {
      * @param directSql
      *      Whether the statement handle is direct SQL (true) or a cursor (false)
      */
-    final void enqueuePreparedStatementDiscardItem(int handle, boolean directSql) {
-        if (this.getConnectionLogger().isLoggable(java.util.logging.Level.FINER))
-            this.getConnectionLogger().finer(this + ": Adding PreparedHandle to queue for un-prepare:" + handle);
+    final void enqueueUnprepareStatementHandle(PreparedStatementHandle statementHandle) {
+        if (loggerExternal.isLoggable(java.util.logging.Level.FINER))
+            loggerExternal.finer(this + ": Adding PreparedHandle to queue for un-prepare:" + statementHandle.getHandle());
 
         // Add the new handle to the discarding queue and find out current # enqueued.
-        this.discardedPreparedStatementHandles.add(new PreparedStatementDiscardItem(handle, directSql));
-        this.discardedPreparedStatementHandleQueueCount.incrementAndGet();
+        discardedPreparedStatementHandles.add(statementHandle);
+        if (0 < discardedPreparedStatementHandleCount.incrementAndGet())
```

Comment:
I noticed when writing the eviction tests that we cannot run the un-prepare batch call while another statement is being prepared so I move this call to handlePreparedStatementDiscardActions to the closePreparedHandle.